app/controllers/github_repos_controller.rb

```diff
@@ -5,77 +5,65 @@ class GithubReposController < ApplicationController
+    @repos = fetch_repositories_from_github(known_repositories_ids)
   rescue Octokit::Unauthorized => e
-    render json: { error: "Github Unauthorized: #{e.message}", status: 401 }, status: :unauthorized
-  end
-
-  def create
-    authorize GithubRepo
-    @repo = GithubRepo.find_or_create(fetched_repo_params(fetch_repo))
-    current_user.touch(:github_repos_updated_at)
-    if @repo.valid?
-      redirect_to "/settings/integrations", notice: "GitHub repo added"
-    else
-      redirect_to "/settings/integrations",
-                  error: "There was an error adding your Github repo"
-    end
-  end
-
-  def update
-    @repo = GithubRepo.find(params[:id])
-    current_user.touch(:github_repos_updated_at)
-    authorize @repo
-    if @repo.update(featured: false)
-      redirect_to "/settings/integrations", notice: "GitHub repo added"
-    else
-      redirect_to "/settings/integrations",
-                  error: "There was an error removing your Github repo"
-    end
+    render json: { error: "GitHub Unauthorized: #{e.message}", status: 401 }, status: :unauthorized
   end
 
   def update_or_create
     authorize GithubRepo
 
     params[:github_repo] = JSON.parse(params[:github_repo])
-    fetched_repo = fetch_repo
+
+    fetched_repo = fetch_repository_from_github(repo_params[:github_id_code])
     unless fetched_repo
-      render json: "error: Could not find Github repo", status: :not_found
+      render json: { error: "GitHub repository not found", status: 404 }, status: :not_found
       return
     end
 
-    repo = GithubRepo.find_or_create(fetched_repo_params(fetched_repo))
+    repo = GithubRepo.upsert(current_user, fetched_repo_params(fetched_repo))
 
     current_user.touch(:github_repos_updated_at)
 
     if repo.valid?
       render json: { featured: repo.featured }
     else
-      render json: "error: #{repo.errors.full_messages}"
+      render json: { error: repo.errors.full_messages, status: 422 }, status: :unprocessable_entity
     end
   end
 
   private
 
+  # TODO: use Github::UserClient or something
   def create_octokit_client
     current_user_token = current_user.identities.where(provider: "github").last.token
-    client = Octokit::Client.new(access_token: current_user_token)
-    client&.repositories&.sort_by!(&:name)
-    client
+    Octokit::Client.new(access_token: current_user_token)
+  end
+
+  def fetch_repositories_from_github(known_repositories_ids)
+    client = create_octokit_client
+
+    client.repositories(visibility: :public).map do |repo|
+      repo.featured = known_repositories_ids.include?(repo.id)
+      repo
+    end.sort_by(&:name)
+  end
+
+  def fetch_repository_from_github(repository_id)
+    client = create_octokit_client
+
+    client.repository(repository_id)
+  rescue Octokit::NotFound
+    nil
   end
 
   def fetched_repo_params(fetched_repo)
     {
       github_id_code: fetched_repo.id,
-      user_id: current_user.id,
```

Comment:
Do we not need this anymore?

Reply:
Nope, we add it in the `GithubRepo.upsert` method